From 968c5acbcec3f5444a1aff3180396400829cbc31 Mon Sep 17 00:00:00 2001
From: Mike Lundy <mike@fluffypenguin.org>
Date: Sat, 19 Dec 2009 04:34:38 -0800
Subject: [PATCH] Add a predicate option to XMonad.Prompt

This allows one to, for example, use isInfixOf instead of isPrefixOf.
---
 XMonad/Prompt.hs        |    5 ++++-
 XMonad/Prompt/Window.hs |    3 +--
 2 files changed, 5 insertions(+), 3 deletions(-)

diff --git a/XMonad/Prompt.hs b/XMonad/Prompt.hs
index 6abb4d0..3dab64a 100644
--- a/XMonad/Prompt.hs
+++ b/XMonad/Prompt.hs
@@ -137,6 +137,7 @@ data XPConfig =
         , autoComplete      :: Maybe Int  -- ^ Just x: if only one completion remains, auto-select it,
         , showCompletionOnTab :: Bool     -- ^ Only show list of completions when Tab was pressed
                                          --   and delay by x microseconds
+        , searchPredicate   :: String -> String -> Bool -- Filter predicate for doPrompt
         }
 
 data XPType = forall p . XPrompt p => XPT p
@@ -210,7 +211,9 @@ defaultXPConfig =
         , historyFilter     = id
         , defaultText       = []
         , autoComplete      = Nothing
-        , showCompletionOnTab = False }
+        , showCompletionOnTab = False
+        , searchPredicate   = isPrefixOf
+        }
 greenXPConfig = defaultXPConfig { fgColor = "green", bgColor = "black" }
 amberXPConfig = defaultXPConfig { fgColor = "#ca8f2d", bgColor = "black", fgHLight = "#eaaf4c" }
 
diff --git a/XMonad/Prompt/Window.hs b/XMonad/Prompt/Window.hs
index 0f0ed2a..a5b6aeb 100644
--- a/XMonad/Prompt/Window.hs
+++ b/XMonad/Prompt/Window.hs
@@ -24,7 +24,6 @@ module XMonad.Prompt.Window
     ) where
 
 import qualified Data.Map as M
-import Data.List
 
 import qualified XMonad.StackSet as W
 import XMonad
@@ -89,7 +88,7 @@ doPrompt t c = do
       bringAction      = winAction bringWindow
       bringCopyAction  = winAction bringCopyWindow
 
-      compList m s = return . filter (isPrefixOf s) . map fst . M.toList $ m
+      compList m s = return . filter (searchPredicate c s) . map fst . M.toList $ m
 
 
 -- | Brings a copy of the specified window into the current workspace.
-- 
1.6.6.rc3

